### PR TITLE
[ENH] RAND Index metric for segmentations

### PR DESCRIPTION
**Pull request by gavinkatz001**
#### Reference Issues/PRs

Fixed the first half of this issue: https://github.com/sktime/sktime/issues/7511 by implementing a RAND Index class that inherits from BaseDetectionMetric (PR for the implementation of this class can be found at https://github.com/sktime/sktime/pull/7515/files). Optimized to O(N) for both space and time complexity.  

#### What does this implement/fix? Explain your changes.

This class implements the computation of the classic RAND Index. It analyzes the similarity between ground truth and predicted segments, returning a float value between 0.0 and 1.0 (0.0 being no similarity, 1.0 being perfect overlap). 

#### Does your contribution introduce a new dependency? If yes, which one?

The only dependency of this new class is that the RandIndex class inherits from BaseDetectionMetric.

#### What should a reviewer concentrate their feedback on?

Please let me know if there are any mistakes in my implementation of the algorithm or if there is any room for further optimization when it comes to computational/spatial efficiency. 

#### Did you add any tests for the change?

Not yet. Would like some guidance on how to write appropriate tests for this class.

#### Any other comments?

The branch associated with this PR (add-f1-and-rand-index) was branched directly off the PR that is still pending approval for the implementation of the base class: https://github.com/sktime/sktime/pull/7515/files. This means that this PR will contain commits from that PR as well. Please let me know if the commits I made for _**this**_  PR regarding the addition of the _randindex.py file and modification of the __init__.py file at sktime\sktime\performance_metrics\detection are for some reason not visible. 

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/performance_metrics/detection/_randindex.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        )
+
+        # 2) Create label arrays: ground_truth_labels[i] = cluster_id of i in ground-truth  # noqa: E501
+        #                         predicted_labels[i]    = cluster_id of i in predicted
+        ground_truth_labels = [None] * total_length
+        predicted_labels = [None] * total_length
+
+        # Assign cluster IDs for ground truth
+        # We can just use an incremental ID for each segment
+        for cluster_id, (start, end) in enumerate(ground_truth_segments):
```
Comment (fkiraly) [inline sktime/performance_metrics/detection/_randindex.py:53]: The segmentation can be labelled, in wihch case the data frame would contain a column `labels`. This assumes it is unlabelled?

File: sktime/performance_metrics/detection/_randindex.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        # 1) Determine total length (N)
+        total_length = max(
+            max(end for _, end in ground_truth_segments),
+            max(end for _, end in predicted_segments),
+        )
+
+        # 2) Create label arrays: ground_truth_labels[i] = cluster_id of i in ground-truth  # noqa: E501
+        #                         predicted_labels[i]    = cluster_id of i in predicted
+        ground_truth_labels = [None] * total_length
```
Comment (fkiraly) [inline sktime/performance_metrics/detection/_randindex.py:48]: this is superfluous - you could get the labels by index from the `transform` method already.

However, doing this is wasteful of memory and compute - the result can be obtained linear in the number of segments.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Good start!

This is halfway there imo, comments:

* the metric converts the segments into the "dense" format returned by `transform` internally. This is superfluous, but also unnecessary - the calculation needs only starts and ends.
* the metric does not take into account that a given segment label could occur multiple times, this should be done - but perhaps better in a second step.
* the metric also uses `iloc` rather than `loc` indexing. If `X` is passed, I believe it should do `loc` indexing? Or do we always want `iloc` indexing?
* there is a faster algorithm than looping over the dense format - just use starts and ends, and compute the size of the rectangles.

**Comment by gavinkatz001** [comment]
changes have been incorporated.

**Comment by fkiraly** [comment]
> changes have been incorporated. let me know what you think and if there is any further feedback that can be added to this iteration.

I have merged `main` into the PR to avoid divergence of branches, to make your life easier - so you can use the `BaseDetectionMericx` - and to avoid divergence of branches.

You need to ensure that you pull all changes instead of overwriting them or reverting them.

**Comment by fkiraly** [comment]
I will have to dive into the code to check what is happening - whether the failures are coming from the class or the framework. Could be either, given how new things are.

**Comment by gavinkatz001** [comment]
Took a few tries but merge issues seem to have been fixed in init.py file.

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think the failure is due to you assuming a different output format for segmentation?

The currently used format is `pd.DataFrame` with an `ilocs` column, that contains `pandas` left-closed segments.

I have added some examples for this data container, for testing, in this PR: https://github.com/sktime/sktime/pull/7546
(in `sktime.detection._datatypes._examples`, you can just call the example generator functions)

**Comment by fkiraly** [comment]
I changed the input conversion to what is expected, this should work now.

As said, you need to assume the output format of `predict`.

**Comment by gavinkatz001** [comment]
Hi Franz, happy new year! Now that the input conversion has been changed, can this PR be merged?

**Comment by fkiraly** [comment]
Happy new your to you too!

Yes, I think that might have fixed some issues. You should be mindful though of keeping your fork updated, or your branch gets outdated and may incur merge conflicts.

**Comment by gavinkatz001** [comment]
Whenever I open up my laptop to develop I always follow the sktime guidelines for developing a new feature (fetching from upstream, merging, etc). Apologies if something was out of sync this time.

**Comment by fkiraly** [comment]
no problem at all, I updated it for you

**Review by fkiraly — CHANGES_REQUESTED** [review]
The main failure seems to be coming from different assumptions in the input format.

In the current convention, we use a single `pandas.Interval` column (called `ilocs`) for the segment starts/ends, your code assumes two columns `start` and `end`.

**Comment by gavinkatz001** [comment]
Hopefully all checks pass with the update for the start/end issue.

**Comment by gavinkatz001** [comment]
Seems that the only failed checks have something to do with a scipy import error. Any idea how to address this?

**Comment by fkiraly** [comment]
> Seems that the only failed checks have something to do with a scipy import error. @fkiraly any idea how to address this?

These are unrelated, the metric looks fine now, interface-wise!

**Comment by gavinkatz001** [comment]
> > Seems that the only failed checks have something to do with a scipy import error. @fkiraly any idea how to address this?
> 
> These are unrelated, the metric looks fine now, interface-wise!

great!

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

A small thing, I think the metric currently computes the rand index in units of `iloc`, in reference eto `X`, not in `loc`.

I think there should be at least an option to compute `loc` if `X` is passed, possibly it should be the default.

Beyond this, really great work on finding an O(n) algorithm!

I think `sorted` is not necessary since we can assume sorted, but I suppose it will only cost O(n) if it is.

**Comment by gavinkatz001** [comment]
> Great!
> 
> A small thing, I think the metric currently computes the rand index in units of `iloc`, in reference eto `X`, not in `loc`.
> 
> I think there should be at least an option to compute `loc` if `X` is passed, possibly it should be the default.
> 
> Beyond this, really great work on finding an O(n) algorithm!
> 
> I think `sorted` is not necessary since we can assume sorted, but I suppose it will only cost O(n) if it is.

just added a fix for this.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

* Could you ensure both settings are also tested, by using `get_test_params`, and having one setting being True, the other False?
* the `use_loc` parameter should appear in a `Parameters` section in the docstring, like in most other objects in the code base.

**Comment by gavinkatz001** [comment]
done :)

**Comment by gavinkatz001** [comment]
> done

getting this error: ValueError: Segment {'start': 0, 'end': 2, 'label': 0} references label(s) not found in X.index: 0 or 2. I ran it through chatgpt o1 and got the below response. Let me know if you think this makes sense:

Analysis

The error

ValueError: Segment {'start': 0, 'end': 2, 'label': 0} references label(s) not found in X.index: 0 or 2
tells you that your RandIndex code tried to interpret start=0 and end=2 as if they were labels in X.index, but X.index is [1, 2, 4, 5, 42, 43]. Consequently, your code raised a ValueError because it couldn’t find 0 or 2 among those index labels.

This happens because:

Your default is use_loc=True.
If X is passed in, and use_loc=True, the code attempts to map each segment’s [start, end] to integer positions via

start_iloc = index_map[start_val]
end_iloc   = index_map[end_val]
where index_map is built from X.index.
If start_val or end_val isn’t actually in X.index, it triggers the ValueError.
But in your test:


y_true = pd.DataFrame({"ilocs": [0, 2, 3]})
X      = pd.DataFrame({"foo": [8, 4, 3, 7, 10, 12]}, index=[1, 2, 4, 5, 42, 43])
the user’s segments for y_true (like [0..1) or [2..3)) do not match the index [1, 2, 4, 5, 42, 43]. The code tries to interpret them as “labels” in X.index, fails to find 0, and crashes.

How to Fix or Work Around It
You have two main approaches:

1. Turn Off use_loc When Passing X
If your segments are truly 0-based integer positions that do not correspond to X.index, you probably don’t want to interpret them as loc-labels. Instead, do:

metric = RandIndex(use_loc=False)
loss   = metric(y_true, y_pred, X)  # now it won't try to map segments to X.index
This ensures “start”/“end” remains an integer-based approach, ignoring X.index.

2. Make Sure Segments Are Actually in X.index
If you do want to interpret “start”/“end” as the real labels from X.index, you must ensure the DataFrame’s start or ilocs columns match labels that exist in X.index. For example:

# If X.index = [1, 2, 4, 5, 42, 43], then "start"/"end" must be in that set:
y_true = pd.DataFrame({"ilocs": [1, 2, 4]})

or, if you’re using the 'start'/'end' columns:

y_true = pd.DataFrame({"start": [1, 4], "end": [4, 5]})
and so on. Then your code can successfully map those labels to integer positions.

Bottom Line
Why it happens: By default, use_loc=True, so your metric tries to convert [start,end] into integer positions by looking up X.index. But 0 is missing in X.index=[1,2,4,5,42,43].
Easy fix: Turn off loc-based mapping (i.e. use_loc=False) if your data is actually 0-based integers and not meant to match X.index.
Alternatively: Provide segments that are valid labels in X.index whenever you call the metric with X and use_loc=True.

**Comment by fkiraly** [comment]
can you explain why you are converting `loc` to `iloc`?

From my understanding of the algorithm, this should never be necessary, because segments are encoded in `iloc`.

**Comment by gavinkatz001** [comment]
> can you explain why you are converting `loc` to `iloc`?
> 
> From my understanding of the algorithm, this should never be necessary, because segments are encoded in `iloc`.

Here is the response:


Why add a “loc-to-iloc” conversion when segments are presumably in iloc already?

By default, one might indeed store breakpoints/segments in 0-based integer positions, matching Python’s array indexing (what you called “iloc”). However, in some use cases—and in many time-series libraries—users may prefer specifying segments by the actual index labels of a pandas DataFrame (what you might call “loc”).
	•	For example, if ￼ has a non-integer index like [“2020-01-01”, “2020-01-02”, …], or simply a numeric but non-sequential index [10, 20, 30, …], then a user might say “a segment starts at label 10, ends at label 20.”
	•	Internally, though, the Rand Index implementation (or any adjacency-based approach) typically operates on 0-based integer positions, i.e. it loops from 0..N to compare pairwise overlaps.

So the loc-to-iloc logic is there to handle both possibilities:
	1.	If someone truly has segments in integer-based “iloc” form (0, 1, 2, etc.) or uses a column ilocs with integer positions, no conversion is needed—the code effectively does nothing special.
	2.	If someone supplies “start”/“end” as labels from ￼, or an ilocs column that is really an interval in label space, then that code block converts “label” → “position” so the rest of the algorithm can remain purely integer-based.

In short:
	•	“We never need it” is valid if your entire pipeline strictly uses 0-based indices for segments.
	•	The loc-to-iloc step becomes important for advanced scenarios where your index is not strictly [0..N) but instead something like [10, 11, 42, 1000, …].

Hence the presence of that conversion code: it’s optional functionality to interpret user segments in label space rather than raw integer positions. If your boss’s project always uses integer “iloc” segments anyway, that conversion code effectively won’t come into play—but it remains available for the more general case.

**Comment by fkiraly** [comment]
> If someone supplies “start”/“end” as labels from X.index, or an ilocs column that is really an interval in label space, then that code block converts “label” → “position” so the rest of the algorithm can remain purely integer-based.

I see! So there is perhaps a very simple resolution, then:

With the current API, this is impossible - segment specifications are by default - and only - `iloc`. This decision was made, in fact, to avoid the aforementioned `loc`-to-`iloc` conversions, which can be fiddly and error prone, and would need to be replicated across metrics, estimators, etc.

And, if at some point we would like to introduce this, it should be handled by the boilerplate, and not an individual metric (for which we still would implement only the `iloc` version).

Further, and this is *important* - if it were true that we never need it, why does it get called in the test?

**Comment by gavinkatz001** [comment]
> Great!
> 
> * Could you ensure both settings are also tested, by using `get_test_params`, and having one setting being True, the other False?
> * the `use_loc` parameter should appear in a `Parameters` section in the docstring, like in most other objects in the code base.

It gets called in the test because you previously asked that I test both settings (use_loc true and false). If my understanding is correct (please point out if I am missing something), you requested that an option be available to compute loc if X is passed, which I implemented via the loc -> iloc method. However since it cannot be used, I will remove this. Would you like me to instead implement a fully loc-based system if X is passed and not use iloc at all? If so, then I am somewhat confused since you mentioned that segment specifications are by default - and only - iloc. 

Can you please provide clarification on the following:

- Why do we need to implement an option for loc if metrics should only implement the iloc version of a certain computing process for metrics?
- If we need a loc implementation alongside the iloc one, it should be entirely in terms of loc (no converting back to iloc), correct?
- Why did we need to test both settings if the test suite only has iloc inputs?
- If metrics are only implemented in iloc and all the test inputs are in terms of iloc, then would commit https://github.com/sktime/sktime/pull/7533/commits/5332ef2b9f7aeae6bc87b142a0053608cc4247f3 (which uses iloc only, and passed all the tests except for that import error) suffice for this metric?

**Comment by gavinkatz001** [comment]
No, you are not missing anything. I just misunderstood what was being asked of me. My apologies for the confusion and thank you for taking the time to make those clarifications Franz. This was very helpful and I really appreciate it!

**Comment by gavinkatz001** [comment]
All checks have passed! :)

**Review by fkiraly — APPROVED** [review]
Looks good!

Approved, but some non-blocking comments:

* What would be nice is a quick mathematical description in the docstring, but it is not blocking.
* if labels are already provided, they should be used, instead of the assigned ones

File: sktime/performance_metrics/detection/_randindex.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        -------
+        float
+            Rand Index score between 0.0 and 1.0.
+        """  # noqa: E501
+        # 1) Validate and extract segments (still iloc-based).
+        y_true_segments = self._extract_segments(y_true, "y_true")
+        y_pred_segments = self._extract_segments(y_pred, "y_pred")
+
+        # 2) Assign unique cluster IDs to each segment.
+        y_true_segments = self._assign_unique_ids(y_true_segments, prefix="true")
```
Comment (fkiraly) [inline sktime/performance_metrics/detection/_randindex.py:58]: sometimes the segments already come with labels, in this case both dataframes will have a column `label`.

In this case, can we not just use that column?

**Comment by gavinkatz001** [comment]
I have added the check to use the labels column if it is present, as well as the mathematical description in the doc string

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

Minor doc detail, you need to add newlines around the bullet points, or they will not render correctly.

**Review by fkiraly — DISMISSED** [review]

**Review by fkiraly — DISMISSED** [review]
I added a mathematical description of what is implemented, kindly review.

**Review by fkiraly — APPROVED** [review]
